src/main/java/org/opensearch/security/action/apitokens/ApiTokenAction.java

```diff
@@ -295,4 +354,99 @@ private void sendErrorResponse(RestChannel channel, RestStatus status, String er
         }
     }
 
+    /**
+     *  Validates that the user has the required permissions to create an API token (must be a subset of their own permissions)
+     * */
+    @SuppressWarnings("unchecked")
+    void validateUserPermissions(List<String> clusterPermissions, List<ApiToken.IndexPermission> indexPermissions) {
```

Comment:
So, the token they are requesting is validated to be a subset of the permissions available to them in roles.yml, right?

What is supposed to happen if the permissions in roles.yml are later changed, especially if they are reduced?

Reply:
I think the same problem exists with roles as well. If you stored roles information and a user's roles mappings changed then the tokens would need to be invalidated.

Reply:
This is a good point. I added a todo on this issue: https://github.com/opensearch-project/security/issues/5088#issuecomment-2637627943

Reply:
I kind of feel that the handling of later privilege changes is under-specified here, as it is actually a quite challenging topic.

Changes of privileges can be caused by different effects:

- Changes in roles.yml
- Changes in roles_mapping.yml
- Change of a user's IP address (roles_mapping.yml can match IP addresses to privileges)
- Changes to the user-to-role-mapping in external authentication systems such as LDAP, OIDC or SAML


Especially the last item makes it impossible to get automatic notifications about the privileges of a user being changed. Thus, IMHO, there needs to be a conceptual review on what can be done and what can't be done and how a compensation for that could look like.